## Deliverable 2. Create a Customer Travel Destinations Map.

In [1]:
# Dependencies and Setup
import pandas as pd
import requests
import gmaps

# Import API key
from config import g_key

# Configure gmaps API key
gmaps.configure(api_key=g_key)

In [2]:
# 1. Import the WeatherPy_database.csv file. 
city_data_df = pd.read_csv("Weather_Database/WeatherPy_database.csv")
city_data_df.head()

City ID         City Country      Lat       Lng  Max Temp  Humidity  \
0        0  Yellowknife      CA  62.4560 -114.3525     55.56        58   
1        1    Jamestown      US  42.0970  -79.2353     74.01        71   
2        2  Tuktoyaktuk      CA  69.4541 -133.0374     43.25        76   
3        3        Airai      TL  -8.9266  125.4092     59.74        92   
4        4   Bredasdorp      ZA -34.5322   20.0403     50.83        87   

   Cloudiness  Wind Speed Current Description  
0         100       13.00     overcast clouds  
1          75       14.00       broken clouds  
2         100       10.09     overcast clouds  
3          99        1.95     overcast clouds  
4          18        6.29          few clouds

In [3]:
# 2. Prompt the user to enter minimum and maximum temperature criteria 

locations = city_data_df[["Lat", "Lng"]]

# Get max temp:

max_temp= city_data_df["Max Temp"]
temps= []

for temp in max_temp:
    temps.append(max(temp, 0))
    
# Assign the figure variable to the gmaps.figure() attribute.  
fig = gmaps.figure(center=(30.0, 31.0), zoom_level=2)

# 4. Assign the heatmap_layer variable to the heatmap_layer attribute and add in the locations.
heat_layer = gmaps.heatmap_layer(locations, weights= [max(temp, 0) for temp in max_temp], dissipating=False, max_intensity=300, point_radius=4)

# 5. Add the heatmap layer.
fig.add_layer(heat_layer)

# 6. Call the figure to plot the data.
fig

Figure(layout=FigureLayout(height='420px'))

In [4]:
# 3. Filter the city_data_df DataFrame using the input statements to create a new DataFrame using the loc method.

# Ask the customer to add a minimum and maximum temp value:

min_temp=float(input("What is the minimum temperature you would like for your trip? "))
max_temp=float(input("What is the maximum temperature you would like for your trip? "))

What is the minimum temperature you would like for your trip? 75
What is the maximum temperature you would like for your trip? 90


In [5]:
# Filter the dataset to find the cities that fit this criteria:

preferred_cities_df=city_data_df.loc[(city_data_df["Max Temp"] <= max_temp) & 
                                     (city_data_df["Max Temp"]>=min_temp)]
preferred_cities_df.head(10)

City ID                  City Country      Lat       Lng  Max Temp  \
6         6             Maryville      US  35.7565  -83.9705     85.30   
15       15                  Hilo      US  19.7297 -155.0900     79.29   
16       16              Sterling      US  39.0062  -77.4286     86.25   
17       17               Tautira      PF -17.7333 -149.1500     77.40   
18       18              Maragogi      BR  -9.0122  -35.2225     77.79   
29       29            Koutsouras      GR  35.0339   25.9469     77.11   
30       30               Qandala      SO  11.4720   49.8728     88.61   
31       31            Tres Picos      MX  15.8667  -93.5333     86.43   
33       33  Vila Franca do Campo      PT  37.7167  -25.4333     77.41   
35       35            Butaritari      KI   3.0707  172.7902     82.22   

    Humidity  Cloudiness  Wind Speed Current Description  
6         58          72        5.99       broken clouds  
15        96          80        5.01       broken clouds  
16        52          75        7.00       broken clouds  
17        75          13       19.95          light rain  
18        69           5       10.27           clear sky  
29        61           0        8.48           clear sky  
30        58          42        1.23    scattered clouds  
31        70         100        4.38     overcast clouds  
33        82          23        4.21          few clouds  
35        66          79       16.44       broken clouds

In [7]:
# 4a. Determine if there are any empty rows.

# Check the number of rows that have a NULL value:
preferred_cities_df.isnull().sum()

City ID                0
City                   0
Country                0
Lat                    0
Lng                    0
Max Temp               0
Humidity               0
Cloudiness             0
Wind Speed             0
Current Description    0
dtype: int64

### Notice that there are no missing rows

In [10]:
# Check total number of rows:
preferred_cities_df.count()

City ID                229
City                   229
Country                229
Lat                    229
Lng                    229
Max Temp               229
Humidity               229
Cloudiness             229
Wind Speed             229
Current Description    229
dtype: int64

### Notice that there are 229 missing rows in all

In [9]:
# 4b. Drop any empty rows and create a new DataFrame that doesn’t have empty rows.

preferred_cities_df.dropna()

City ID          City Country      Lat       Lng  Max Temp  Humidity  \
6          6     Maryville      US  35.7565  -83.9705     85.30        58   
15        15          Hilo      US  19.7297 -155.0900     79.29        96   
16        16      Sterling      US  39.0062  -77.4286     86.25        52   
17        17       Tautira      PF -17.7333 -149.1500     77.40        75   
18        18      Maragogi      BR  -9.0122  -35.2225     77.79        69   
..       ...           ...     ...      ...       ...       ...       ...   
657      657  Chinna Salem      IN  11.6500   78.9000     80.58        71   
658      658   Brazzaville      CG  -4.2658   15.2832     84.83        51   
661      661   Kloulklubed      PW   7.0419  134.2556     83.32        71   
672      672       Caruray      PH  10.3094  119.0128     80.13        85   
673      673        Panabá      MX  21.2833  -88.2667     87.87        63   

     Cloudiness  Wind Speed Current Description  
6            72        5.99       broken clouds  
15           80        5.01       broken clouds  
16           75        7.00       broken clouds  
17           13       19.95          light rain  
18            5       10.27           clear sky  
..          ...         ...                 ...  
657         100        1.88     overcast clouds  
658           8        3.22           clear sky  
661         100        7.27          light rain  
672         100        9.93     overcast clouds  
673          16       10.18          few clouds  

[229 rows x 10 columns]

### Notice the number of rows after dropna() is also 229. This is just reconfirmation that there were no NULL values

------

In [14]:
# 5a. Create DataFrame called hotel_df to store hotel names along with city, country, max temp, and coordinates.
hotel_df = preferred_cities_df[["City", "Country", "Max Temp", "Current Description", "Lat", "Lng"]].copy()

# 5b. Create a new column "Hotel Name"
hotel_df["Hotel Name"] = ""
hotel_df.head()

City Country  Max Temp Current Description      Lat       Lng  \
6   Maryville      US     85.30       broken clouds  35.7565  -83.9705   
15       Hilo      US     79.29       broken clouds  19.7297 -155.0900   
16   Sterling      US     86.25       broken clouds  39.0062  -77.4286   
17    Tautira      PF     77.40          light rain -17.7333 -149.1500   
18   Maragogi      BR     77.79           clear sky  -9.0122  -35.2225   

   Hotel Name  
6              
15             
16             
17             
18

### Retrieve Hotels from Nearby Search

In [39]:
# 6a. Set parameters to search for hotels with 5000 meters.
params = {
    "radius": 5000,
    "type": "lodging",
    "key": g_key
}

In [40]:
# 6b. Iterate through the hotel DataFrame.
for index, row in hotel_df.iterrows():
    
    # 6c. Get latitude and longitude from DataFrame:
    
    lat = row["Lat"]
    lng = row["Lng"] 


    # Add the latitude and longitude to location key for the params dictionary.
    params["location"] = f"{lat},{lng}"

    # 6d. Set up the base URL for the Google Directions API to get JSON data.
    
    base_url = "https://maps.googleapis.com/maps/api/place/nearbysearch/json"

    # 6e. Make request and retrieve the JSON data from the search. 
   
    hotels = requests.get(base_url, params=params).json()
    
    # 6f. Get the first hotel from the results and store the name, if a hotel isn't found skip the city.
    
    try:
        hotel_df.loc[index, "Hotel Name"] = hotels["results"][0]["name"]
    except (IndexError):
        print("Hotel not found... skipping.")    

Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.


In [41]:
# 7. Drop the rows where there is no Hotel Name

hotel_df.head()

City Country  Max Temp Current Description      Lat       Lng  \
6   Maryville      US     85.30       broken clouds  35.7565  -83.9705   
15       Hilo      US     79.29       broken clouds  19.7297 -155.0900   
16   Sterling      US     86.25       broken clouds  39.0062  -77.4286   
17    Tautira      PF     77.40          light rain -17.7333 -149.1500   
18   Maragogi      BR     77.79           clear sky  -9.0122  -35.2225   

                                           Hotel Name  
6   Holiday Inn Express & Suites Alcoa (Knoxville ...  
15                                Hilo Hawaiian Hotel  
16  Hampton Inn & Suites Washington-Dulles Interna...  
17                                       Pension Oaoa  
18           Salinas do Maragogi All Inclusive Resort

In [42]:
# 8a. Create the output File (CSV):

output_data_file = "Vacation_Search/WeatherPy_vacation.csv"

# 8b. Export the City_Data into a csv:

hotel_df.to_csv(output_data_file, index_label="City_ID")

In [45]:
# 9. Using the template add city name, the country code, the weather description and maximum temperature for the city:

info_box_template = """
<dl>
<dt>Hotel Name</dt><dd>{Hotel Name}</dd>
<dt>City</dt><dd>{City}</dd>
<dt>Country</dt><dd>{Country}</dd>
<dt>Current Weather Description</dt><dd>{Current Description}</dd>
<dt>Max Temp</dt><dd>{Max Temp} °F</dd>
</dl>
"""

# 10a. Get the data from each row and add it to the formatting template and store the data in a list:

hotel_info = [info_box_template.format(**row) for index, row in hotel_df.iterrows()]

# 10b. Get the latitude and longitude from each row and store in a new DataFrame:

locations = hotel_df[["Lat", "Lng"]]

# 11a. Add a marker layer for each city to the map:

max_temp = hotel_df["Max Temp"]
fig = gmaps.figure(center=(30.0, 31.0), zoom_level=1.5)
heat_layer = gmaps.heatmap_layer(locations, weights=max_temp, dissipating=False,
             max_intensity=300, point_radius=4)
marker_layer = gmaps.marker_layer(locations, info_box_content=hotel_info)
fig.add_layer(heat_layer)
fig.add_layer(marker_layer)

# 11b. Display the figure:

fig

Figure(layout=FigureLayout(height='420px'))